src/app/filecoin-plus/allocators/components/SelectSort.tsx

```diff
@@ -0,0 +1,78 @@
+'use client'
+
+import {
+  Listbox as HeadlessUIListbox,
+  ListboxButton as HeadlessUIListboxButton,
+  ListboxOptions as HeadlessUIListboxOptions,
+  ListboxOption as HeadlessUIListboxOption,
+} from '@headlessui/react'
+import { ArrowsDownUp, CaretDown, Check } from '@phosphor-icons/react/dist/ssr'
+import type { Column, SortDirection } from '@tanstack/react-table'
+
+import { Icon } from '@/components/Icon'
+
+import type { Allocator } from '../schemas/AllocatorSchema'
+
+export type TableSortOption = {
+  id: SortDirection
+  name: string
+}
+
+type SortFilterProps = {
+  column: Column<Allocator>
+  options: ReadonlyArray<TableSortOption>
+  defaultOption: TableSortOption
+}
+
+export function SelectSort({
+  column,
+  options,
+  defaultOption,
+}: SortFilterProps) {
+  const currentSortId = column.getIsSorted()
+
+  const selectedOption =
+    options.find((option) => option.id === currentSortId) || defaultOption
+
+  return (
+    <HeadlessUIListbox value={selectedOption} onChange={handleSortChange}>
+      <HeadlessUIListboxButton className="inline-flex w-full items-center justify-between gap-2 rounded-lg border border-brand-300 bg-brand-800 p-3 text-brand-300 focus:brand-outline hover:border-current hover:text-brand-400">
+        <div className="inline-flex items-center gap-2">
+          <Icon component={ArrowsDownUp} />
+          <span className="hidden text-nowrap md:block">
+            {selectedOption.name}
+          </span>
+        </div>
+        <span className="hidden md:block">
+          <Icon component={CaretDown} size={16} weight="bold" />
+        </span>
+      </HeadlessUIListboxButton>
+
+      <HeadlessUIListboxOptions
+        as="ul"
+        anchor={{ to: 'bottom end', gap: 12 }}
+        className="rounded-lg border border-brand-100 bg-brand-800 py-2 text-brand-100 focus:brand-outline focus-within:outline-2"
+      >
+        {options.map((option) => (
+          <HeadlessUIListboxOption
+            key={option.id}
+            as="li"
+            value={option}
+            className="group flex cursor-default items-center justify-between gap-12 text-nowrap bg-transparent px-5 py-2 data-[focus]:bg-brand-500"
+          >
+            <span>{option.name}</span>
+            <span className="invisible mb-px group-data-[selected]:visible">
+              <Icon component={Check} size={20} />
+            </span>
+          </HeadlessUIListboxOption>
+        ))}
+      </HeadlessUIListboxOptions>
+    </HeadlessUIListbox>
+  )
+
+  function handleSortChange(newOption: TableSortOption) {
+    if (newOption.id !== currentSortId) {
+      column.toggleSorting()
+    }
+  }
```

Comment:
This implementation of `toggleSorting()` might not be intended. You can reproduce it by trying A-Z, Z-A, and then A-Z again - The latter, you'll have to hit twice.

The problem seem to be that `column.toggleSorting()` without any arguments cycles through the sorting states: asc, desc, and none. This means:

- If the current state is asc and newOption.id is desc, calling toggleSorting() once will set it to desc, but another call would set it to none.
- Because we’re not explicitly passing a direction, it often takes multiple clicks to reach the desired state, especially when switching between asc and desc.

So, this version leads to unpredictable behavior where it can take multiple clicks to reach the intended sort order. It effectively relies on the component’s internal state and cycles through it, which is inconsistent.

In the suggested version, column.toggleSorting(newOption.id === 'desc') directly sets the sort direction based on newOption.id:

- Passing newOption.id === 'desc' tells toggleSorting exactly which direction to sort (ascending or descending) without cycling through states.
- When newOption.id is 'asc', it sorts in ascending order; when it’s 'desc', it sorts in descending order immediately, eliminating the need for multiple clicks.

```suggestion
    function handleSortChange(newOption: TableSortOption) {
      column.toggleSorting(newOption.id === 'desc')
    }
```

Reply:
You are absolutely right @mirhamasala, great catch! I don't know why I didn't catch it... Anyway, fixed it!